pubsub/pstest/fake_test.go

```diff
@@ -219,6 +278,138 @@ func TestSubscriptionErrors(t *testing.T) {
 	checkCode(err, codes.NotFound)
 }
 
+func TestSubscriptionDeadletter(t *testing.T) {
+	_, _, server, cleanup := newFake(context.TODO(), t)
+	defer cleanup()
+
+	ctx := context.Background()
+
+	topic, err := server.GServer.CreateTopic(ctx, &pb.Topic{
+		Name: "projects/P/topics/in",
+	})
+	if err != nil {
+		t.Fatalf("failed to create in topic")
+	}
+	deadletterTopic, err := server.GServer.CreateTopic(ctx, &pb.Topic{
+		Name: "projects/P/topics/deadletter",
+	})
+	if err != nil {
+		t.Fatalf("failed to create deadletter topic")
+	}
+	retries := 3
+	sub, err := server.GServer.CreateSubscription(ctx, &pb.Subscription{
+		Name:               "projects/P/subscriptions/S",
+		Topic:              topic.Name,
+		AckDeadlineSeconds: 10,
+		DeadLetterPolicy: &pb.DeadLetterPolicy{
+			DeadLetterTopic:     deadletterTopic.Name,
+			MaxDeliveryAttempts: int32(retries),
+		},
+	})
+	if err != nil {
+		t.Fatalf("failed to create subscription")
+	}
+	dlSub, err := server.GServer.CreateSubscription(ctx, &pb.Subscription{
+		Name:               "projects/P/subscriptions/SD",
+		Topic:              deadletterTopic.Name,
+		AckDeadlineSeconds: 10,
+	})
+	if err != nil {
+		t.Fatalf("failed to create subscription")
+	}
+
+	messageData := []byte("message data")
+	_, err = server.GServer.Publish(ctx, &pb.PublishRequest{
+		Topic: topic.Name,
+		Messages: []*pb.PubsubMessage{
+			{
+				Data: messageData,
+			},
+		},
+	})
+	if err != nil {
+		t.Fatalf("failed to publish message")
+	}
+	rand.Seed(time.Now().UTC().UnixNano())
+	maxAttempts := rand.Intn(5) + retries
+	for i := 0; i < maxAttempts; i++ {
+		pull, err := server.GServer.Pull(ctx, &pb.PullRequest{
+			Subscription: sub.Name,
+			MaxMessages:  10,
+		})
+		if err != nil {
+			t.Fatalf("failed during pull")
+		}
+		if i < retries {
+			if len(pull.ReceivedMessages) != 1 {
+				t.Fatalf("expected 1 message received a different number %d", len(pull.ReceivedMessages))
+			}
```

Comment:
After adding the fix above to sync message.DeliveryAttempts, add verification for `m.proto.GetDeliveryAttempts()` here.

Reply:
Check [fake_test.go#349](https://github.com/googleapis/google-cloud-go/pull/5320/files#diff-c6a74920cef99f2ab6281118b9eb61b39c99f9b17ee2961758db6688e0c6af9eR349)